packages/datadog-plugin-fs/src/index.js

```diff
@@ -0,0 +1,488 @@
+    const tags = makeFSTags(resourceName, fd, null, config, tracer)
+    if (typeof uid === 'number') {
+      tags['file.uid'] = uid.toString()
+    }
+    if (typeof gid === 'number') {
+      tags['file.gid'] = gid.toString()
+    }
+    return tags
+  }
+}
+
+function createFchownTags (resourceName, config, tracer) {
+  return function fchownTags (fd, uid, gid) {
+    if (typeof this === 'object' && this !== null && this.fd) {
+      gid = uid
+      uid = fd
+      fd = this.fd
+    }
+    const tags = makeFSTags(resourceName, fd, null, config, tracer)
+    if (typeof uid === 'number') {
+      tags['file.uid'] = uid.toString()
+    }
+    if (typeof gid === 'number') {
+      tags['file.gid'] = gid.toString()
+    }
+    return tags
+  }
+}
+
+function getSymbolName (sym) {
+  return sym.description || sym.toString()
+}
+
+function hasParent (tracer) {
+  return !!tracer.scope().active()
+}
+
+function createWrapCb (tracer, config, name, tagMaker) {
+  const makeTags = tagMaker(name, config, tracer)
+  name = 'fs.' + name
+  return function wrapFunction (fn) {
+    return tracer.wrap('fs.operation', function () {
+      if (typeof arguments[arguments.length - 1] !== 'function') {
+        return
+      }
+      const tags = makeTags.apply(this, arguments)
+      return tags ? { tags } : null
+    }, fn, true)
+  }
+}
+
+function createWrap (tracer, config, name, tagMaker) {
+  const makeTags = tagMaker(name, config, tracer)
+  return function wrapSyncFunction (fn) {
+    return tracer.wrap('fs.operation', function () {
+      const tags = makeTags.apply(this, arguments)
+      return tags ? { tags } : null
+    }, fn, true)
+  }
+}
+
+function makeFSFlagTags (resourceName, path, options, defaultFlag, config, tracer) {
+  const tags = makeFSTags(resourceName, path, options, config, tracer)
+
+  if (tags) {
+    let flag = defaultFlag
+    if (typeof options === 'object' && options !== null) {
+      if (options.flag) {
+        flag = options.flag
+      } else if (options.flags) {
+        flag = options.flags
+      }
+    }
+    tags['file.flag'] = flag
+    return tags
+  }
+}
+
+function makeFSTags (resourceName, path, options, config, tracer) {
+  path = options && 'fd' in options ? options.fd : path
+  const tags = {
+    'component': 'fs',
+    'resource.name': resourceName,
+    'service.name': config.service || `${tracer._service}-fs`
+  }
+
+  switch (typeof path) {
+    case 'object': {
+      if (path === null) return
+      const src = 'src' in path ? path.src : null
+      const dest = 'dest' in path ? path.dest : null
+      if (src || dest) {
+        tags['file.src'] = src
+        tags['file.dest'] = dest
+      } else {
+        tags['file.path'] = path
+      }
+      break
+    }
+    case 'string': {
+      tags['file.path'] = path
+      break
+    }
+    case 'number': {
+      tags['file.descriptor'] = path.toString()
+      break
+    }
+  }
+
+  return tags
+}
+
+function getFileHandlePrototype (fs) {
```

Comment:
Is this guaranteed to run before any method is called or referenced?

Reply:
It only gets called in `patch`, so I think so?

Reply:
What I mean is that since it returns a promise, it's possible that the unpatched versions have been called by the time the promise resolves.

Reply:
Unfortunately the only way to get this object is through that promise call. The class isn't exposed synchronously. That being said, as long as we're required first, the first call to `fs.promises.open()` will be ours, so the only way we won't be the first code with access to the prototype is if somehow it takes longer to get a filehandle than in user code. I don't think that's likely to happen. If `UV_THREADPOOL_SIZE` is set to 1, then I don't think it's even possible.